### ISSUE TITLE
ICE: "global closures can't capture anything"

### ISSUE THREAD

**Issue by aidanhs** [rust-lang/rust#37550]
```
#![feature(const_fn)]
fn main() {}
const fn x() {
    let t = true;
    let x = || t;
}
```
https://is.gd/JFNOLu

```
rustc 1.14.0-nightly (3f4408347 2016-10-27)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'global closures can't capture anything', ../src/librustc_passes/consts.rs:587
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

(const fn tracking issue https://github.com/rust-lang/rust/issues/24111)

**Comment by eddyb** [comment]
This sounds like a pass ordering issue, MIR check passes should come first (which would error about the variables).

**Comment by Mark-Simulacrum** [comment]
Appears to be fixed on `rustc 1.16.0-nightly (468227129 2017-01-03)`.

```
error[E0016]: blocks in constant functions are limited to items and tail expressions
 --> test.rs:4:13
  |
4 |     let t = true;
  |             ^^^^

error[E0016]: blocks in constant functions are limited to items and tail expressions
 --> test.rs:5:13
  |
5 |     let x = || t;
  |             ^^^^

error: aborting due to 2 previous errors
```

**Comment by Mark-Simulacrum** [comment]
E-needstest as per previous comment.
